src/main/java/com/github/dariobalinzo/elastic/ElasticRepository.java

```diff
@@ -35,70 +38,122 @@
+        final PointInTimeBuilder pitBuilder;
+        try {
+            pitBuilder = new PointInTimeBuilder(Optional.ofNullable(cursor.pitId()).orElse(openPit(cursor.index())));
+            pitBuilder.setKeepAlive("20m");
+
+            final var searchSourceBuilder = new SearchSourceBuilder()
+                    .query(queryBuilder)
+                    .pointInTimeBuilder(pitBuilder)
+                    .size(pageSize);
+
+            cursor.cursorFields().forEach(cursorField -> searchSourceBuilder.sort(cursorField.field(), SortOrder.ASC));
+            Optional.ofNullable(cursor.sortValues()).ifPresent(searchSourceBuilder::searchAfter);
+
+            final var searchRequest = new SearchRequest().source(searchSourceBuilder);
+            final var response = executeSearch(searchRequest);
+
+            final var documents = extractDocuments(response);
+            final var totalHits = response.getHits().getTotalHits().value;
+
+            final PageResult result;
+            if (documents.isEmpty() || totalHits == 0) {
+                // return empty page with same cursor to maybe try again later
+                result = PageResult.empty(cursor);
+            } else if (totalHits > cursor.runningDocumentCount() + documents.size()) {
+                // return page with scrollable cursor
+                // note: ES says the pitId can change in a query so it's got to be set each query cycle
+                final var scrollable = cursor.scrollable(pitBuilder.getEncodedId(),
+                        response.getHits().getAt(documents.size() - 1).getSortValues(), documents.size(),
+                        totalHits);
+
+                result = PageResult.intermediatePage(documents, scrollable);
+            } else {
+                // return last page and cursor for the next frame
+                final var reframed = cursor.reframe(response.getHits().getAt(documents.size() - 1).getSortValues());
+                result = PageResult.lastPage(documents, reframed);
+            }
+
+            if (result.lastPage()) {
+                closePit(pitBuilder.getEncodedId());
+            }
 
-        List<Map<String, Object>> documents = extractDocuments(response);
+            return result;
 
-        Cursor lastCursor;
-        if (documents.isEmpty()) {
-            lastCursor = Cursor.empty();
-        } else {
-            Map<String, Object> lastDocument = documents.get(documents.size() - 1);
-            lastCursor = new Cursor(lastDocument.get(cursorField).toString());
+        } catch (IOException | InterruptedException e) {
+            throw new RuntimeException(e);
+        } catch (ElasticsearchStatusException e) {
+            if (cursor.isScrollable()) {
+                try {
+                    closePit(cursor.pitId());
+                } catch (RuntimeException ioException) {
+                    // nothing, best efforts, probably here because pit was closed unexpectedly anyway (i.e. timed out)
```

Comment:
narrow the exception window here. to broad - will catch anything and everything, need to check cause probably